### PR TITLE
integrate datetime_generator and get_ordered_col

### PR DESCRIPTION
**Pull request by drahc1R**
Replacement to PR https://github.com/capitalone/synthetic-data/pull/318
Integrated get_ordered_columns into dataset_generator.
Reimplemented datetime_generator to work with get_ordered_column.
Implemented unit and integration tests.

### CODE REVIEW

File: tests/test_dataset_generator.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 from unittest import mock
 
+import dataprofiler as dp
 import numpy as np
 import pandas as pd
 from numpy.random import PCG64, Generator
 
-from synthetic_data.dataset_generator import generate_dataset
+from synthetic_data import dataset_generator as dg
+from synthetic_data.distinct_generators import datetime_generator as dategen
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:12]: alias could be a bit more descriptive variable name

File: tests/test_dataset_generator.py [hunk 2]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+        self.start_date = pd.Timestamp(2001, 12, 22)
+        self.end_date = pd.Timestamp(2023, 1, 1)
+        self.date_format_list = ["%B %d %Y %H:%M:%S"]
+
+    def test_get_ordered_column_datetime_ascending(self):
+        data = dategen.random_datetimes(
+            rng=self.rng, start_date=self.start_date, end_date=self.end_date, num_rows=5
+        )
+
+        ordered_data = np.array(
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:118]:
```suggestion
        expected = np.array(
```

Comment (taylorfturner) [reply tests/test_dataset_generator.py:118]: will need to update references to this too

Comment (taylorfturner) [reply tests/test_dataset_generator.py:118]: `expected` and `actual` make it super clear what they are doing

File: tests/test_dataset_generator.py [hunk 3]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+                    "September 27 2018 18:24:03",
+                    datetime.strptime(
+                        "September 27 2018 18:24:03", self.date_format_list[0]
+                    ),
+                ],
+            ]
+        )
+
+        ordered_data = ordered_data[:, 0]
+        output_data = dg.get_ordered_column(data, "datetime", "ascending")
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:154]:
```suggestion
        actual = dg.get_ordered_column(data, "datetime", "ascending")
```

File: tests/test_dataset_generator.py [hunk 4]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+
+        np.testing.assert_array_equal(output_data, ordered_data)
+
+    def test_get_ordered_column_datetime_descending(self):
+        data = dategen.random_datetimes(
+            rng=self.rng, start_date=self.start_date, end_date=self.end_date, num_rows=5
+        )
+
+        ordered_data = np.array(
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:163]:
```suggestion
        expected = np.array(
```

File: tests/test_dataset_generator.py [hunk 5]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+                    "October 02 2006 22:34:32",
+                    datetime.strptime(
+                        "October 02 2006 22:34:32", self.date_format_list[0]
+                    ),
+                ],
+            ]
+        )
+
+        ordered_data = ordered_data[:, 0]
+        output_data = dg.get_ordered_column(data, "datetime", "descending")
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:199]:
```suggestion
        actual = dg.get_ordered_column(data, "datetime", "descending")
```

File: tests/test_dataset_generator.py [hunk 6]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+                [
+                    "September 27 2018 18:24:03",
+                    datetime.strptime(
+                        "September 27 2018 18:24:03", self.date_format_list[0]
+                    ),
+                ],
+            ]
+        )
+
+        ordered_data = ordered_data[:, 0]
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:153]: probably don't need this.... can just call `ordered_data[:,0]` in `assert_array_equal`

File: tests/test_dataset_generator.py [hunk 7]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+        output_data = dg.get_ordered_column(data, "datetime", "descending")
+
+        np.testing.assert_array_equal(output_data, ordered_data)
+
+    def test_get_ordered_column_custom_datetime_ascending(self):
+        custom_date_format = ["%Y %m %d"]
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:204]: we should probably have multiple values here ... to validate it processes correctly if it is more than one value in the list

File: tests/test_dataset_generator.py [hunk 8]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+                    "March 11 2016 15:15:39",
+                    "September 27 2018 18:24:03",
+                ]
+            ),
+        ]
+        ordered_data = np.array(ordered_data, dtype=object)
+
+        output_data = []
+        for data_type in data.keys():
+            output_data.append(dg.get_ordered_column(data[data_type], data_type))
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:347]: why not call `generate_dataset`?

File: tests/test_dataset_generator.py [hunk 9]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+                            datetime.strptime(
+                                "October 02 2006 22:34:32", self.date_format_list[0]
+                            ),
+                        ],
+                    ]
+                ),
+            }
+        )
+
+        ordered_data = [
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:328]: expected?

File: tests/test_dataset_generator.py [hunk 10]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+        )
+
+        ordered_data = ordered_data[:, 0]
+        output_data = dg.get_ordered_column(data, "datetime", "descending")
+
+        np.testing.assert_array_equal(output_data, ordered_data)
+
+    def test_get_ordered_column(self):
+
+        data = OrderedDict(
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:285]: expected?

File: tests/test_dataset_generator.py [hunk 11]
```diff
@@ -91,8 +95,256 @@ def test_generate_custom_dataset(self):
...
+                ],
+            ]
+        )
+
+        ordered_data = ordered_data[:, 0]
+        output_data = dg.get_ordered_column(data, "datetime", "descending")
+
+        np.testing.assert_array_equal(output_data, ordered_data)
+
+    def test_get_ordered_column(self):
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:283]: add to `TestDatasetGenerator`

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: tests/test_dataset_generator.py [hunk 12]
```diff
@@ -1,11 +1,17 @@
 import unittest
+from collections import OrderedDict
+from datetime import datetime
 from unittest import mock
 
+pass
+
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:7]: remove

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: tests/distinct_generators/test_datetime_generator.py [hunk 13]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
                 try:
-                    pd.to_datetime(date_str, format=date_format)
+                    pd.to_datetime(date[1], format=date[0])
                     format_success[i] = True
                 except ValueError:
                     pass
         self.assertGreater(sum(format_success), 1)
+
+    def test_random_datetimes_output(self):
+        outputs = [
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py:90]: expected

File: tests/distinct_generators/test_datetime_generator.py [hunk 14]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
+            np.testing.assert_array_equal(self.random_datetimes_output[i], outputs[i])
+
+    def test_generate_datetime_output(self):
+        self.assertTrue(self.generate_datetime_output[0] == "2006 10 02")
+        self.assertTrue(
+            self.generate_datetime_output[1]
+            == datetime.strptime(
+                self.generate_datetime_output[0], self.date_format_list[0]
+            )
+        )
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py]: move to L35 function and rename L35 function

File: tests/distinct_generators/test_datetime_generator.py [hunk 15]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
+        dates = date_generator.random_datetimes(
+            self.rng, start_date=self.start_date, end_date=self.end_date
         )
-        for date_str in result:
+        for date in dates:
             try:
-                pd.to_datetime(date_str, format="%B %d %Y %H:%M:%S")
+                pd.to_datetime(date, format="%B %d %Y %H:%M:%S")
             except ValueError:
                 self.fail("pd.to_datetime() raised ValueError for default formatting")
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py:75]: not sure this is needed

File: tests/distinct_generators/test_datetime_generator.py [hunk 16]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
+        format_success = [False] * len(date_formats)
+        for date in self.random_datetimes_output:
+            for i in range(len(date_formats)):
                 try:
-                    pd.to_datetime(date_str, format=date_format)
+                    pd.to_datetime(date[1], format=date[0])
                     format_success[i] = True
                 except ValueError:
                     pass
         self.assertGreater(sum(format_success), 1)
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py:87]: not sure needed

File: tests/distinct_generators/test_datetime_generator.py [hunk 17]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
         self.assertTrue(date_obj <= self.end_date)
 
     def test_random_datetimes_return_type_and_size(self):
-        result = date_generator.random_datetimes(
-            self.rng, self.date_format_list, self.start_date, self.end_date, 5
-        )
-        self.assertIsInstance(result, np.ndarray)
-        self.assertEqual(result.shape[0], 5)
+        self.assertIsInstance(self.random_datetimes_output, np.ndarray)
+        self.assertEqual(self.random_datetimes_output.shape[0], 10)
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py:65]: move to L35 function

File: tests/distinct_generators/test_datetime_generator.py [hunk 18]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
 
     def test_generate_datetime_range(self):
-        date_str = date_generator.generate_datetime(
-            self.rng, self.date_format_list[0], self.start_date, self.end_date
+        date_obj = pd.to_datetime(
+            self.generate_datetime_output[1], format=self.generate_datetime_output[0]
         )
-        date_obj = pd.to_datetime(date_str, format=self.date_format_list[0])
         self.assertTrue(self.start_date <= date_obj)
         self.assertTrue(date_obj <= self.end_date)
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py:61]: L35 function?

File: tests/distinct_generators/test_datetime_generator.py [hunk 19]
```diff
@@ -26,55 +33,73 @@ def test_start_end_date_when_none(self):
...
-            self.rng, self.date_format_list[0], self.start_date, self.end_date
-        )
         try:
-            pd.to_datetime(date_str, format=self.date_format_list[0])
+            pd.to_datetime(
+                self.generate_datetime_output[1],
+                format=self.generate_datetime_output[0],
+            )
         except ValueError:
             self.fail("pd.to_datetime() raised ValueError for custom formatting")
```
Comment (taylorfturner) [inline tests/distinct_generators/test_datetime_generator.py:54]: needed?

File: tests/test_dataset_generator.py [hunk 20]
```diff
@@ -43,21 +47,147 @@ def setUp(self):
...
+            ),
+            np.array(["Y", "Y", "Y", "Y", "Y", "Y", "Z", "Z", "Z", "Z"]),
+            np.array(
+                [3.035, 3.477, 4.234, 4.812, 4.977, 5.131, 5.379, 5.488, 7.318, 7.4]
+            ),
+        ]
+        expected_df = pd.DataFrame.from_dict(
+            dict(zip(["int", "dat", "txt", "cat", "flo"], expected_data))
+        )
+        df = dataset_generator.generate_dataset(
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:118]:
```suggestion
        actual_df = dataset_generator.generate_dataset(
```

File: tests/test_dataset_generator.py (hunk 11 above)
Comment (taylorfturner) [reply tests/test_dataset_generator.py:390]: remove this entirely

**Review by taylorfturner — CHANGES_REQUESTED** [review]

**Review by micdavis — DISMISSED** [review]

File: tests/test_dataset_generator.py [hunk 21]
```diff
@@ -91,8 +224,168 @@ def test_generate_custom_dataset(self):
...
+        custom_date_format = ["%Y %m %d", "%B %d %Y %H:%M:%S"]
+        data = datetime_generator.random_datetimes(
+            rng=self.rng,
+            date_format_list=custom_date_format,
+            start_date=self.start_date,
+            end_date=self.end_date,
+            num_rows=5,
+        )
+
+        expected = np.array(
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:340]: it looks like the other tests in `TestGetOrderedColumn` are trying to test `get_ordered_column` which returns an array of strings... but the expected values defined in the test aren't arrays of strings...

Comment (taylorfturner) [reply tests/test_dataset_generator.py:340]: the other tests have the expected of `[:,0]` here but not this test when the test are only changing from `ascending` to `descending`? Do we need `datetime.strptime("2018 09 27", custom_date_format[0]),` values in the `expected` array defined on 340?

File: tests/test_dataset_generator.py [hunk 22]
```diff
@@ -43,25 +48,153 @@ def setUp(self):
...
+        expected_df = pd.DataFrame.from_dict(
+            dict(zip(["int", "dat", "txt", "cat", "flo"], expected_data))
+        )
+        actual_df = dataset_generator.generate_dataset(
+            self.rng,
+            columns_to_generate=columns_to_gen,
+            dataset_length=self.dataset_length,
+        )
+
+        self.assertTrue(actual_df.equals(expected_df))
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:125]: just for consistency and readability.... I would change this to `np.testing.assert_array_equals(actual_df.values, expected_df.values)`

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: tests/test_dataset_generator.py [hunk 23]
```diff
@@ -43,25 +48,152 @@ def setUp(self):
...
+        ]
+        expected_df = pd.DataFrame.from_dict(
+            dict(zip(["int", "dat", "txt", "cat", "flo"], expected_data))
+        )
+        actual_df = dataset_generator.generate_dataset(
+            self.rng,
+            columns_to_generate=columns_to_gen,
+            dataset_length=self.dataset_length,
+        )
+        np.testing.assert_array_equal(actual_df, expected_df)
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:124]:
```suggestion
        np.testing.assert_array_equal(actual_df.values, expected_df.values)
```

File: tests/test_dataset_generator.py [hunk 24]
```diff
@@ -1,11 +1,16 @@
 import unittest
+
+pass
+pass
```
Comment (taylorfturner) [inline tests/test_dataset_generator.py:4]: remove

**Review by taylorfturner — CHANGES_REQUESTED** [review]

**Review by taylorfturner — APPROVED** [review]

**Review by micdavis — APPROVED** [review]

**Pull request merged**